src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs

```diff
@@ -0,0 +1,226 @@
+        : base (initialSignalCount, maximumSignalCount, spinCount, onWait)
+    {
+        CreateAsyncWait(maximumSignalCount);
+    }
+
+#pragma warning disable IDE0060
+    private void CreateAsyncWait(int maximumSignalCount)
+#pragma warning restore IDE0060
+    {
+        lifo_semaphore = InitInternal(2);
+    }
+
+    [MethodImplAttribute(MethodImplOptions.InternalCall)]
+    private static extern void DeleteInternal(IntPtr semaphore);
+
+    public void Dispose()
+    {
+        DeleteInternal(lifo_semaphore);
+        lifo_semaphore = IntPtr.Zero;
+    }
+
+    [MethodImplAttribute(MethodImplOptions.InternalCall)]
+    private static extern void ReleaseInternal(IntPtr semaphore, int count);
+
+    protected override void ReleaseCore(int count)
+    {
+        ReleaseInternal(lifo_semaphore, count);
+    }
+
+    private sealed record WaitEntry (LowLevelLifoAsyncWaitSemaphore Semaphore, int TimeoutMs, Action<LowLevelLifoAsyncWaitSemaphore, object?> OnSuccess, Action<LowLevelLifoAsyncWaitSemaphore, object?> OnTimeout, object? State);
+
+    public void PrepareAsyncWait(int timeoutMs, Action<LowLevelLifoAsyncWaitSemaphore, object?> onSuccess, Action<LowLevelLifoAsyncWaitSemaphore, object?> onTimeout, object? state)
+    {
+        //FIXME(ak): the async wait never spins. Shoudl we spin a little?
+        Debug.Assert(timeoutMs >= -1);
+
+        // Try to acquire the semaphore or
+        // [[a) register as a spinner if false and timeoutMs > 0]]
+        // b) register as a waiter if [[there's already too many spinners or]] true and timeoutMs > 0
+        // c) bail out if timeoutMs == 0 and return false
+        Counts counts = _separated._counts;
+        while (true)
+        {
+            Debug.Assert(counts.SignalCount <= _maximumSignalCount);
+            Counts newCounts = counts;
+            if (counts.SignalCount != 0)
+            {
+                newCounts.DecrementSignalCount();
+            }
+            else if (timeoutMs != 0)
+            {
+                // Maximum number of spinners reached, register as a waiter instead
+                newCounts.IncrementWaiterCount();
+            }
+
+            Counts countsBeforeUpdate = _separated._counts.InterlockedCompareExchange(newCounts, counts);
+            if (countsBeforeUpdate == counts)
+            {
+                if (counts.SignalCount != 0)
+                {
+                    onSuccess (this, state);
+                    return;
+                }
+                if (newCounts.WaiterCount != counts.WaiterCount)
+                {
+                    PrepareAsyncWaitForSignal(timeoutMs, onSuccess, onTimeout, state);
+                    return;
+                }
+                if (timeoutMs == 0)
+                {
+                    onTimeout (this, state);
+                    return;
+                }
+                break;
+            }
+
+            counts = countsBeforeUpdate;
+        }
+
+        Debug.Fail("unreachable");
+
+#if false
```

Comment:
why is this here?

Reply:
I was hoping someone would tell me that it's a good idea to keep the spin-waiting code. I'll delete it.

Reply:
I guess whether to spin-wait or not would depend on perf and other considerations. For instance, is it better to yield back to the JS event loop immediately, or is it better and is it ok to spin-wait a bit for more work and only yield back after the spin-wait fails (potentially the spin-wait may succeed successively and not yield back to the JS event loop for a long time).

Reply:
> potentially the spin-wait may succeed successively and not yield back to the JS event loop for a long time

I think this is a good reason to hold off on adding spin-waiting here until we see how real code behaves. Potentially starving JS wouldn't be great

Reply:
I think you can leave this code here for now but not enable it, because the risk of starving the main thread is big yeah